feat(udf): Add deno as UDF language

bakjos opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR add the option to use deno as udf language, it includes `fetch`, `web` and `crypto` API support. 

```sql
CREATE FUNCTION digest( t string ) RETURNS bytea LANGUAGE deno AS $$
    const subtle = crypto.subtle; 
    const key = await subtle.generateKey({
        name: 'HMAC',
        hash: 'SHA-256',
        length: 256,
        }, true, ['sign', 'verify']);
    
        const enc = new TextEncoder();
        const message = enc.encode(t);
    
        const result = await subtle.sign({
        name: 'HMAC',
        }, key, message);
        console.log('result', result);
        return result;
$$ SET function_type = 'async';
```

```sql
CREATE FUNCTION delay_response()
    RETURNS TABLE (x int) LANGUAGE deno AS $$
                 const delayedResponses = {
                    delays: [50, 10, 15], 
                    wait(delay) {
                      return new Promise((resolve) => {
                        setTimeout(resolve, delay);
                      });
                    },
                  
                    async *[Symbol.asyncIterator]() {
                      for (const delay of this.delays) {
                        await this.wait(delay);
                        yield delay;
                      }
                    },
                  };
                return delayedResponses;
$$   SET function_type = 'normal';
```

```sql
CREATE FUNCTION fetch_api() RETURNS TABLE ( data struct< id int, name string, gender string >) LANGUAGE deno AS $$
    const response = await fetch('https://rickandmortyapi.com/api/character');
    const resp = await response.json();
    for (const r of resp.results) {
        yield r;
    }
$$ SET function_type = 'async_generator';
```

I added support for setting a parameter when creating the function to define the function_type: 

| type | function | declaration |
| ---- |  ----- | ------ |
| `async` | A java script async function, to enable the usage of the await api | ` async function <name> { <body> } ` | 
| `async_generator` | A java script async generator function, to enable the usage of the await and yield api | ` async function* <name> { <body> } ` | 
| `generator` | A java script generator function, to enable the usage of the yield api | ` function* <name> { <body> } ` | 
| `normal` | A regular javascript function, if a promise is returned, will wait until is fulfilled to return the result | `function <name> { <body> } ` | 

If not defined will use `normal` for UDFs and `generator` for table functions;

I put the deno language support under a feature flag, which is currently enabled by default. (Plan to remove before this PR gets merged).

It will require for https://github.com/risingwavelabs/arrow-udf/pull/7 to be merged 

<!--

**Please do not leave this empty!**

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)
- Refer to a related PR or issue link (optional)

-->

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)
<!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->
- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.


<!--
Please create a release note for your changes.

Discuss technical details in the "What's changed" section, and
focus on the impact on users in the release note.

You should also mention the environment or conditions where the impact may occur.
-->

</details>

xxchan reviewed (commented):
Thanks for your amazing work! Could you share the background and your motivation of the work? like why do you need it, what's its advantage over current `js` UDF, what's not achievable now? 

Users may feel strange that there are both `js` UDF and `deno` UDF, if they are not very familiar with js runtimes. So it's important to figure out whether we want both of them, and if so, make clear the difference for users.

bakjos commented:
@xxchan https://risingwave-community.slack.com/archives/C02T3F7UYM6/p1709825596494219

xxchan commented:
Got it. Sorry that I missed the Slack discussion. For such large feature, it would still be nicer if we open an issue to have larger visibility and more discussion before the PR. ❤️

bakjos commented:
Also do you think the sintax `CREATE [TABLE|SOURCE ] .. AS FUNCTION <function_name> (param1, param2)` could work?

wangrunji0408 commented:
> I added support for setting a parameter when creating the function to define the function_type:

According to postgres' documentation, the `SET param TO value` is to override a specific system parameter during the function call. Using this syntax to define properties of the function itself seems like a misuse.

For simplicity, is it possible to just assume `async function` for scalar functions and `async function*` for table functions? Would performance be hurt if adding `async` declaration to normal js functions? Is there some logic that can not be expressed in this situation?

wangrunji0408 commented on src/frontend/src/handler/create_function.rs:
```diff
@@ -50,7 +50,7 @@ pub async fn handle_create_function(
             let lang = lang.real_value().to_lowercase();
             match &*lang {
-                "python" | "java" | "wasm" | "rust" | "javascript" => lang,
+                "python" | "java" | "wasm" | "rust" | "javascript" | "deno" => lang,
```
I don't think we should introduce `deno` as a new language. From the user's perspective, they are just writing javascript. Normally they shouldn't care which backend the function will run on. But since now we have 2 backends available, how about introducing a new option `RUNTIME [quickjs | deno]` for setting this?

bakjos replied:
Sounds good

wangrunji0408 commented on src/frontend/src/handler/create_function.rs:
```diff
@@ -186,6 +188,41 @@ pub async fn handle_create_function(
+                (None, Some(_as)) => body = Some(_as.into_string()),
+                (Some(CreateFunctionUsing::Link(link)), None) => {
+                    let bytes = download_code_from_link(&link).await?;
+                    compressed_binary = Some(zstd::stream::encode_all(bytes.as_slice(), 0)?);
```
I don't think javascript source code needs to be compressed and stored in `compressed_binary`. Perhaps storing the plain code in `body` is enough.

bakjos replied:
This is for the cases where you want to load compiled javascript (as for the sse library)

wangrunji0408 replied:
But is the compiled bytecode stable across different V8 version? If not, I'm afraid the compiled code won't run on a newer runtime in the future.

wangrunji0408 replied:
Got it. That sounds reasonable.

wangrunji0408 commented on src/frontend/src/handler/create_function.rs:
```diff
@@ -186,6 +188,41 @@ pub async fn handle_create_function(
+                    let bytes = download_code_from_link(&link).await?;
+                    compressed_binary = Some(zstd::stream::encode_all(bytes.as_slice(), 0)?);
+                }
+                (Some(CreateFunctionUsing::Base64(encoded)), None) => {
```
Similarly, base64 encoding seems unnecessary for the source code.

wangrunji0408 commented on src/sqlparser/src/ast/mod.rs:
It would be better not to hardcode parameters in the parser, but to use general identifier as parameter instead.

wangrunji0408 commented:
Please also add some end-to-end tests in `e2e_test/udf/js_deno_udf.slt`. You can use the `system` command in sqllogictest to start a mock server.

https://github.com/risingwavelabs/risingwave/blob/fb7fadb906890dd8268198a1a814797054f243de/e2e_test/udf/retry_python.slt#L1-L2

bakjos commented on src/connector/src/sink/encoder/json.rs:
```diff
@@ -855,6 +855,8 @@ mod tests {
             ["schema"]
             .to_string();
         let ans = r#"{"fields":[{"field":"v1","optional":true,"type":"boolean"},{"field":"v2","optional":true,"type":"int16"},{"field":"v3","optional":true,"type":"int32"},{"field":"v4","optional":true,"type":"float"},{"field":"v5","optional":true,"type":"string"},{"field":"v6","optional":true,"type":"int32"},{"field":"v7","optional":true,"type":"string"},{"field":"v8","optional":true,"type":"int ...
+        let schema = serde_json::from_str::<Value>(&schema).unwrap();
+        let ans = serde_json::from_str::<Value>(ans).unwrap();
         assert_eq!(schema, ans);
```
It was due an update in the dependencies, the order of the string changed, so I changed it to use json instead of just plain string

bakjos replied:
This was because `deno_core` adds the feature `preserve_order` for `serde_json`, I [patched](https://github.com/bakjos/deno_core/commit/70d1544ed07380d3f37a92118d5ca4f1f20c6616)  the dependency to remove it, but I'm not sure if this is the right place to do so https://github.com/risingwavelabs/risingwave/blob/cd2c0d4cb00e18f727653586c6977a6e8fda267f/Cargo.toml#L309-L310

wangrunji0408 replied:
I'm okay with this patch. cc @xxchan any better idea on this feature interaction?

bakjos commented:
> > I added support for setting a parameter when creating the function to define the function_type:
> 
> According to postgres' documentation, the `SET param TO value` is to override a specific system parameter during the function call. Using this syntax to define properties of the function itself seems like a misuse.
> 
> For simplicity, is it possible to just assume `async function` for scalar functions and `async function*` for table functions? It seems possible because I see that all functions will be converted to promises in `arrow-udf-js-deno`. However, if explicit `async` configuration can not be avoided, I'd prefer to introduce a new `ASYNC` keyword.

I like the idea of introducing new keywords, the main issue is with [AsyncIterators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncIterator) vs generators  in the table function since it changes the signature of the function, so `NORMAL` `ASYNC` and `ASYNC GENERATOR` are needed

wangrunji0408 commented on e2e_test/udf/sse/server/src/main.rs:
Is it possible to rewrite this in Python like `mock_server.py`? This way the test pipeline can be simplified.

bakjos replied:
I tried initially using python but couldn't find a good library to mock the SSE server

bakjos replied:
I ended up using flask and removing the rust version

wangrunji0408 commented on Cargo.toml:
```diff
@@ -306,6 +307,8 @@ futures-timer = { git = "https://github.com/madsim-rs/futures-timer.git", rev =
 etcd-client = { git = "https://github.com/risingwavelabs/etcd-client.git", rev = "4e84d40" }
 # todo(wcy-fdu): remove this patch fork after opendal release a new version to apply azure workload identity change.
 reqsign = { git = "https://github.com/wcy-fdu/reqsign.git", rev = "e6cb304" }
+# patch to remove preserve_order from serde_json
+deno_core = { git = "https://github.com/bakjos/deno_core", rev = "9b241c6" }
```
Is it due to feature conflicts with RisingWave itself?

wangrunji0408 reviewed (commented):
Sorry for the late response as I was on vacation last week. This PR generally LGTM now. Thanks for your patient work!

bakjos commented on src/expr/core/src/expr/expr_udf.rs:
```diff
@@ -152,6 +158,10 @@ impl UserDefinedFunction {
             UdfImpl::Python(runtime) => runtime.call(&self.identifier, &arrow_input),
+            #[cfg(feature = "embedded-deno-udf")]
+            UdfImpl::Deno(runtime) => {
+                futures::executor::block_on(runtime.call(&self.identifier, arrow_input))
```
I also tried this, but it starts complaining about 

```
error[E0277]: `(dyn std::any::Any + 'static)` cannot be sent between threads safely
   --> src/expr/core/src/expr/expr_udf.rs:104:65
    |
104 |       async fn eval_row(&self, input: &OwnedRow) -> Result<Datum> {
    |  _________________________________________________________________^
105 | |         let mut columns = Vec::with_capacity(self.children.len());
106 | |         for child in &self.children {
107 | |             let datum = child.eval_row(input).await?;
...   |
113 | |         Ok(output_array.to_datum())
114 | |     }
    | |_____^ `(dyn std::any::Any + 'static)` cannot be sent between threads safely
````

bakjos replied:
I changed it to use `block_in_place + block_on` to avoid blocking the streaming worker

bakjos commented on src/meta/dashboard/src/proxy.rs:
```diff
@@ -87,8 +88,17 @@ async fn proxy(
     let resp = CachedResponse {
-        code: content.status(),
-        headers: content.headers().clone(),
+        code: hyper::StatusCode::from_u16(content.status().as_u16()).unwrap(),
+        headers: content.headers().iter().fold(
+            hyper::HeaderMap::with_capacity(content.headers().len()),
+            |mut header, it| {
+                header.insert(
+                    hyper::header::HeaderName::from_str(it.0.as_str()).unwrap(),
+                    hyper::header::HeaderValue::from_bytes(it.1.as_bytes()).unwrap(),
+                );
+                header
+            },
+        ),
         body: content.bytes().await?,
```
This is needed because I had to downgrade `reqwest` to 0.11 to match the `deno_http` version (due to the recent axum version update)

bakjos replied:
I ended up patching deno to use request 0.12 instead of changing this file

wangrunji0408 replied:
But it seems to bring more submodules even if the feature is not enabled. 😕
```
+ cargo build -p risingwave_cmd_all --profile dev --features rw-static-link
    Updating git repository `https://github.com/bakjos/deno`
    Updating git submodule `https://github.com/denoland/deno_std`
    Updating git submodule `https://github.com/web-platform-tests/wpt.git`
    Updating git submodule `https://github.com/denoland/node_test.git`
```

BugenZhao replied:
> I had to downgrade `reqwest` to 0.11 to match the `deno_http` version (due to the recent axum version update)

Can we add `reqwest 0.11` a dependency under a different name? Different versions should work together without issues.

wangrunji0408 replied:
It's due to a version conflict of an inner dependency. 😕

```
error: failed to select a version for `hyper-util`.
    ... required by package `deno_http v0.141.0`
    ... which satisfies dependency `deno_http = "^0.141.0"` of package `arrow-udf-js-deno-runtime v0.0.1 (https://github.com/risingwavelabs/arrow-udf.git?rev=4c4e176#4c4e1762)`
    ... which satisfies git dependency `arrow-udf-js-deno-runtime` (locked to 0.0.1) of package `arrow-udf-js-deno v0.0.1 (https://github.com/risingwavelabs/arrow-udf.git?rev=4c4e176#4c4e1762)`
    ... which satisfies git dependency `arrow-udf-js-deno` (locked to 0.0.1) of package `risingwave_expr v1.7.0-alpha (/Users/wangrunji/Codes/risingwave/src/expr/core)`
    ... which satisfies path dependency `risingwave_expr` (locked to 1.7.0-alpha) of package `risingwave_cmd_all v1.7.0-alpha (/Users/wangrunji/Codes/risingwave/src/cmd_all)`
versions that meet the requirements `=0.1.2` are: 0.1.2

all possible versions conflict with previously selected packages.

  previously selected package `hyper-util v0.1.3`
    ... which satisfies dependency `hyper-util = "^0.1.3"` (locked to 0.1.3) of package `reqwest v0.12.2`
    ... which satisfies dependency `reqwest = "^0.12.2"` (locked to 0.12.2) of package `risedev v1.7.0-alpha (/Users/wangrunji/Codes/risingwave/src/risedevtool)`

failed to select a version for `hyper-util` which could resolve this conflict
```

BugenZhao replied:
> It's due to a version conflict of an inner dependency. 😕

Do we have any idea why the upstream pins `hyper-util` with `=`?

bakjos replied:
I think will pull the modules, but not compile them, I also patched the deno dependencies to use request 0.12, if you pull the latest commit, should compile ok

BugenZhao commented on Cargo.lock:
```diff
@@ -3807,9 +4343,9 @@ dependencies = [
 [[package]]
 name = "elliptic-curve"
-version = "0.13.6"
+version = "0.13.8"
```
Are these version bumps expected? Would you mind regenerating the `Cargo.lock` file based on the latest revision on the main branch to keep the changes minimal?

xxchan replied:
Please do this
```
# bring in the current latest file
git checkout main -- Cargo.lock
# run any cargo command to update Cargo.lock
cargo check
```

xxchan commented on proto/catalog.proto:
```diff
@@ -259,6 +259,8 @@ message Function {
   // The zstd-compressed binary of the function.
   optional bytes compressed_binary = 17;
   bool always_retry_on_network_error = 16;
+  optional string runtime = 18;
+  optional string function_type = 19;
```
Please comment the meaning and/or possible values of the fields here. Thanks

xxchan commented on proto/expr.proto:
```diff
@@ -537,6 +537,9 @@ message UserDefinedFunction {
   // - If `language` is `rust` or `wasm`, the zstd-compressed wasm binary.
   optional bytes compressed_binary = 10;
   bool always_retry_on_network_error = 9;
+
+  optional string runtime = 11;
+  optional string function_type = 12;
```
ditto

bakjos replied:
Done

xxchan commented on src/meta/model_v2/migration/src/m20230908_072257_init.rs:
I think we need a new migration for adding columns to a table now, and cannot update the old migration files, since sql meta backend is released. cc @yezizp2012

bakjos replied:
Done

xxchan reviewed (commented):
I don't have large concerns since it's feature gated. Can proceed according to how @wangrunji0408 feels.

bakjos commented:
> Could you help resolve conflict in Cargo.lock? I tried but it left with some problems. Feel free to revert my commit.

Done

yezizp2012 commented on src/meta/model_v2/migration/src/m20240408_142249_runtime.rs:
```diff
@@ -0,0 +1,61 @@
+use sea_orm_migration::prelude::{Table as MigrationTable, *};
```
Should use a better naming for this migration file.

wangrunji0408 commented:
Seems the deno e2e test is still flaky. 👀

```
Caused by:
    query failed: connection closed
    [SQL] select * FROM call_sse();
    at e2e_test/udf/deno_udf.slt:233
```

bakjos commented:
> Seems the deno e2e test is still flaky. 👀
> 
> ```
> Caused by:
>     query failed: connection closed
>     [SQL] select * FROM call_sse();
>     at e2e_test/udf/deno_udf.slt:233
> ```

@wangrunji0408  This started to happen after I switched from rust to python to the mock server, changed the protocol from HTTP 1 to HTTP 1.1 to see if it fixes it

bakjos commented:
@wangrunji0408 Updating the protocol didn't work, so I made some extra changes over https://github.com/risingwavelabs/arrow-udf/pull/12 and changed the tests to use a single mock server for both sse and rest, and it seems to have fixed the issue. I ran the test 100 times and got no crash/error. Merged the latest changes to run the tests twice, and it was also successful

xiangjinwu commented:
hakari disabled by #16142 was added back by commit 9d1f3c8df6f6799593d67879ed9322b4cadc63c4

https://github.com/risingwavelabs/risingwave/pull/15719/files/124b1bbf0f3bb288c1ea6760148d234348ca568c
https://github.com/risingwavelabs/risingwave/pull/15719/files/9d1f3c8df6f6799593d67879ed9322b4cadc63c4

cc @xxchan

xiangjinwu commented:
Also unsure what's wrong here:
https://buildkite.com/risingwavelabs/pull-request/builds/47096#018ecb75-4727-4abc-9a4c-66cdafe0d97c

bakjos commented:
@xiangjinwu This PR did not disable hakari, it was disabled by https://github.com/risingwavelabs/risingwave/pull/16142 and it's still in main https://github.com/risingwavelabs/risingwave/blob/86713399ccd08f14d986391f40d32273268e4fb0/Makefile.toml#L1181-L1182, will check the error once again, it could be related to a dependency
